### ISSUE TITLE
Examples are outdated, do not build due to API changes in version 1.67.0

### ISSUE THREAD

**Issue by Xeverous** [cierelabs/x3_fun#12]
From Spirit changelog:

> Removed with_context PR#239.

Here are links to related changes that break compilation:

- https://github.com/boostorg/spirit/pull/239
- https://github.com/boostorg/spirit/pull/254
- https://github.com/boostorg/spirit/commit/4d2241ee5285f9a8546819b698fa27ff44d6cdc6

**Comment by mjcaisse** [comment]
I haven't looked at this project for some time and I guess I hadn't considered that other people might be using it. Thank you for bringing this to my attention. This issue will result in some changes about how I maintain code that was written for a presentation.

**Comment by Xeverous** [comment]
It still gets attention because main page of Boost Spirit links to this
repository as an example.

**Comment by djowel** [comment]
Perhaps it's best to move them to boost. I'll see what I can do.

**Comment by djowel** [comment]
@Xeverous do you think you can make a PR on the changes required?

**Comment by Xeverous** [comment]
Actually not. I tried building the example with Boost 1.68 and while it compiles (after changing `with_context` to `context`) it does not link. Undefined reference to `bool fun::parser::some_member_func_i_dont_remember<IteratorTemplateSpam>`. 

Source of the problem: in `expression.hpp` `BOOST_SPIRIT_DECLARE(expression_type)` adds some member functions but they are just declarations.

**Comment by m-dhooge** [comment]
Funny how things happen!
I just started yesterday to work again on code I developed earlier in 2018 and since then my version of boost upgraded from 65 to 67… And since my code was heavily inspired by x3_fun (BTW, best example ever ;-) ! ), it is also now broken :-/

**Comment by djowel** [comment]
Let's fix it then, and have it in Boost. I'll see what I can do.

**Comment by m-dhooge** [comment]
The error message is:
```
undefined reference to
bool fun::parser::parse_rule<
  __gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >,
  boost::spirit::x3::context<
    boost::spirit::x3::error_handler_tag,
    std::reference_wrapper<boost::spirit::x3::error_handler<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >,
    boost::spirit::x3::context<
      boost::spirit::x3::skipper_tag,
      boost::spirit::x3::char_class<
        boost::spirit::char_encoding::ascii,
        boost::spirit::x3::space_tag
      > const,
      boost::spirit::x3::unused_type
    >
  >,
  fun::ast::expression
>
(
  boost::spirit::x3::rule<
    fun::parser::expression_class,
    fun::ast::expression,
    false
  >,
  __gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >&,
  __gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > const&,
  boost::spirit::x3::context<
    boost::spirit::x3::error_handler_tag,
    std::reference_wrapper<boost::spirit::x3::error_handler<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >,
    boost::spirit::x3::context<boost::spirit::x3::skipper_tag, boost::spirit::x3::char_class<boost::spirit::char_encoding::ascii, boost::spirit::x3::space_tag> const, boost::spirit::x3::unused_type>
  > const&,
  fun::ast::expression&
)
```

**Comment by m-dhooge** [comment]
When I `objdump -dC expression.o`, I can find a single disassembly of parse_rule but with the `std::reference_wrapper` being `const`.

**Comment by Xeverous** [comment]
Perhaps this from `config.hpp` is a problem?

    typedef x3::context<
        error_handler_tag
      , std::reference_wrapper<error_handler_type> const // <- this?
      , phrase_context_type>
    context_type;

**Comment by m-dhooge** [comment]
Yep! This indeed solves the link error (and make my old code work again -- which is also great from my own point of view).

But this is in contradiction with https://github.com/boostorg/spirit/pull/254 which still has the `const`.
However I didn't try to compile it myself.

**Comment by m-dhooge** [comment]
OK, so I tried to compile calc8 with a very dirty `c++ -o calc *.cpp`.
And it seems to have the same kind of parse_rule problem!

PS: calc7 compiles well…
